File: design_patterns/C++/Behavioral/observer.cxx
#include <iostream>
#include <string>
#include <vector>

class JobPost {
  protected:
	std::string title = "";

  public:
	JobPost(std::string title) {
		this->title = title;
	}

	std::string getTitle() {
		return this->title;
	}
};

class JobSeeker {
  protected:
	std::string name = "";

  public:
	JobSeeker(std::string name) {
		this->name = name;
	}

	void onJobPosted(JobPost* job) {
		std::cout << "Hi " << this->name << "! New job posted: "
		          << job->getTitle() << std::endl;
	}
};

class JobPostings {
  protected:
	std::vector<JobSeeker*> observers;

  public:
	void notify(JobPost* jobPosting) {
		for (auto observer : this->observers) {
			observer->onJobPosted(jobPosting);
		}
	}

	void attach(JobSeeker* observer) {
		this->observers.push_back(observer);
	}

	void addJob(JobPost* jobPosting) {
		this->notify(jobPosting);
	}
};

int main() {
	// create subscribers
	auto john = new JobSeeker("John Doe");
	auto jane = new JobSeeker("Jane Doe");
	// create publisher and attach subscribers
	auto jobPostings = new JobPostings();
	jobPostings->attach(john);
	jobPostings->attach(jane);
	// add a new job and see if subscribers get notified
	jobPostings->addJob(new JobPost("Software Engineer"));
	// ==============================================
	// Hi John Doe! New job posted: Software Engineer
	// Hi Jane Doe! New job posted: Software Engineer
	// ==============================================
	return 0;
}

File: design_patterns/Python/Behavioral/observer.py
class JobPost:

    def __init__(self, title):
        self.title = title

    def get_title(self):
        return self.title


class JobSeeker:

    def __init__(self, name):
        self.name = name

    def on_job_post(self, job):
        print('Hi {}! New job posted: {}'.format(self.name, job.title))


class JobPostings:

    def __init__(self):
        self.observers = list()

    def notify(self, job_posting):
        for observer in self.observers:
            observer.on_job_post(job_posting)

    def attach(self, observer):
        self.observers.append(observer)

    def add_job(self, job_posting):
        self.notify(job_posting)


if __name__ == '__main__':

    # create subscribers
    john_doe = JobSeeker('John Doe')
    jane_doe = JobSeeker('Jane Doe')
    kane_doe = JobSeeker('Kane Doe')

    # create publisher and attach subscribers
    job_posting = JobPostings()
    job_posting.attach(john_doe)
    job_posting.attach(jane_doe)

    # add a new job and see if subscribers get notified
    job_posting.add_job(JobPost('Software Engineer'))

    # output:
    # Hi John Doe! New job posted: Software Engineer
    # Hi Jane Doe! New job posted: Software Engineer
